core/shared/src/main/scala/fs2/concurrent/Channel.scala

```diff
@@ -116,174 +118,179 @@ sealed trait Channel[F[_], A] {
+            )
+          }
 
-        def waitOnBound(producer: Deferred[F, Unit], poll: Poll[F]) =
-          poll(producer.get).onCancel {
-            state.update { s =>
-              s.copy(producers = s.producers.filter(_._2 ne producer))
+          def trySend(a: A): F[Either[Channel.Closed, Boolean]] =
+            isClosed.flatMap { b =>
+              if (b)
+                LeftClosedF.asInstanceOf[F[Either[Channel.Closed, Boolean]]]
+              else
+                q.tryOffer(a.asInstanceOf[AnyRef]).map(_.asRight[Channel.Closed])
             }
-          }
 
-        def signalClosure = closedGate.complete(())
+          val stream: Stream[F, A] = {
+            val takeN: F[Chunk[A]] =
+              q.tryTakeN(None).flatMap {
+                case Nil =>
+                  // if we land here, it either means we're consuming faster than producing
+                  // or it means we're actually closed and we need to shut down
+                  // this is the unhappy path either way
+
+                  val fallback = q.take.flatMap { a =>
+                    // if we get the sentinel, shut down all the things, otherwise emit
+                    if (a eq Sentinel)
+                      closedLatch.complete(()).as(Chunk.empty[A])
+                    else
+                      Chunk.singleton(a.asInstanceOf[A]).pure[F]
+                  }
+
+                  // check to see if we're closed and done processing
+                  // if we're all done, complete the latch and terminate the stream
+                  isQuiesced.flatMap { b =>
+                    if (b)
+                      closedLatch.complete(()).as(Chunk.empty[A])
+                    else
+                      fallback
+                  }
+
+                case as =>
+                  // this is the happy path: we were able to take a chunk
+                  // meaning we're producing as fast or faster than we're consuming
+
+                  isClosed.flatMap { b =>
+                    val back = if (b) {
+                      // if we're closed, we have to check for the sentinel and strip it out
+                      val as2 = as.filter(_ ne Sentinel)
+
+                      // if it's empty, we definitely stripped a sentinel, so just be done
+                      // if it's non-empty, we can't know without expensive comparisons, so fall through
+                      if (as2.isEmpty)
+                        closedLatch.complete(()).as(Chunk.empty[A])
+                      else
+                        Chunk.seq(as2).pure[F]
+                    } else {
+                      Chunk.seq(as).pure[F]
+                    }
+
+                    back.asInstanceOf[F[Chunk[A]]]
+                  }
+              }
 
-        @inline private def shouldEmit(s: State) = s.values.nonEmpty || s.producers.nonEmpty
+            // we will emit non-empty chunks until we see the empty chunk sentinel
```

Comment:
Same as above, the use of empty chunks plus sentinel reads a bit convoluted to me

Reply:
Differently purposes. `Sentinel` is a signal to a blocked `take`. `Chunk.empty` is a signal to the typed `stream`. By stripping `Sentinel` in the taker, we avoid the need to continuously check for it in `stream`. We *do* have to continuously check for `Chunk.empty` but that's a lot more efficient since we don't need to scan the chunks.